### ISSUE TITLE
1.2.0rc1 volumes not updating correctly

### ISSUE THREAD

**Issue by DarkFranX** [docker/swarm#2014]
### Short story:
Volumes are not updated correctly unless you restart the swarm containers.

Specifically, if I modify a text file on host machine by adding (or deleting) characters, the same file in the volume will have space at the end (or will be cropped) by the same amount of characters that I added (or removed).

### Detailed story:

Basically, I set up the swarm with virtualbox and consul discovery according to this tutorial: http://blog.scottlowe.org/2015/03/06/running-own-docker-swarm-cluster/

**Docker engine version:** 1.10.3
**Swarm image version:** 1.2.0rc1

Steps to reproduce:

1. `docker -H=tcp://192.168.33.11:3375 run -v /path/to/file.txt:/container/path/to/file.txt:Z iron/gcc:dev sh -c "cat /container/path/to/file.txt"`
2. Edit file.txt
3. `docker -H=tcp://192.168.33.11:3375 run -v /path/to/file.txt:/container/path/to/file.txt:Z iron/gcc:dev sh -c "cat /container/path/to/file.txt"`

Second command should output wrong data.
That happens whether you communicate with Swarm Manager or directly with a node.

### Example: 
**With file.txt:**
```
 /* Hello World program */

#include<stdio.h>

int main()
{
    printf("Hello World");
}
```

IN:
 `docker -H=tcp://192.168.33.11:3375 run -v /path/to/file.txt:/container/path/to/file.txt:Z iron/gcc:dev sh -c "cat /container/path/to/file.txt"`

OUT: 
```
 /* Hello World program */

#include<stdio.h>

int main()
{
    printf("Hello World");
}
```

**Edit file.txt by removing couple lines at the top and save:**
```
int main()
{
    printf("Hello World");
}
```
**Then retry to bind it in the container:**

IN:
 `docker -H=tcp://192.168.33.11:3375 run -v /path/to/file.txt:/container/path/to/file.txt:Z iron/gcc:dev sh -c "cat /container/path/to/file.txt"`

OUT:
```
/* Hello World program */

#include<stdio.
```

**Comment by dongluochen** [comment]
> That happens whether you communicate with Swarm Manager or directly with a node.

Swarm would run the command on a selected node. If you see the problem directly with a node then it's not a Swarm issue. 

The problem may be in Engine or docker-py. I cannot reproduce this problem on Docker 1.11 Engine with Docker CLI. I notice https://docs.docker.com/engine/userguide/containers/dockervolumes/ 
```
Note: Many tools used to edit files including vi and sed --in-place may result in an inode change. Since Docker v1.1.0, this will produce an error such as “sed: cannot rename ./sedKdJ9Dy: Device or resource busy”. In the case where you want to edit the mounted file, it is often easiest to instead mount the parent directory.
```

**Comment by DarkFranX** [comment]
I was mistaken when I said I was doing it on the node directly. I was actually communicating with the **swarm image on the node** (via 192.168.33.12:2375). 

Additionally, I do not encounter that issue when using docker-engine directly. 

I was aware of the inode problem. I will try to bind the parent directory instead of the file specifically and will report back. Although I **did not have this problem with previous version of the swarm image**.

**Comment by dongluochen** [comment]
What's `docker -H manager:port info` in your set up? Do you have a one node cluster? If you have more than a node, the following command may be run on different nodes, i.e., different files.

```
docker -H=tcp://192.168.33.11:3375 run -v /path/to/file.txt:/container/path/to/file.txt:Z iron/gcc:dev sh -c "cat /container/path/to/file.txt"
```

**Comment by DarkFranX** [comment]
This is the node Manager : -H=tcp://192.168.33.11:3375

My nodes' swarm images are advertising to 192.168.33.10 (consul discovery) and their address are : 
Node1: 192.168.33.12:2375
Node2: 192.168.33.13:2375
Node3: 192.168.33.14:2375
Node4: 192.168.33.15:2375

**Comment by dongluochen** [comment]
@DarkFranX When you run the consecutive commands `docker -H manager:port run -v ...`, are they forwarded to the same Docker node? You can see that if you enable debugging in Swarm log (`swarm -l debug manage ...`). If it is, can you run the commands on that Docker Engine directly to see if it has the same output. If it isn't, you are looking at different files on different nodes. 

I don't see Swarm doing much besides forwarding the requests here. I tried your command on Ubuntu Wily machines and I do not have a repro. I'm not clear if `iron/gcc:dev` does anything to the files. Do you see the same if it changes to ubuntu?

**Comment by DarkFranX** [comment]
I don't know if it is forwarded to the same node and I can't try right now (until next week).

However, let me explain my setup:

Each swarm container is in a ubuntu virtual machine. So my manager is in VM1, node1 in VM2, node2 in VM3, node3 in VM4 and node4 in VM5. The sixth virtual machine has a consul container to discover nodes. 

The file.txt is actually in a shared folder between my host and the virtual machines. **This is not related to the issue since the file is correctly update in the VMs**

If I use the docker engine of those VMs (not the swarm), everything is working has expected: the file is correctly bound to the container.

Previously, I had swarm:latest. **And with the exactly same setup it worked!** Then I moved to swarm:1.2.0rc1 because I needed the --tmpfs functionality and **now file binds do not work as it should. **

Hope this helps, I thought that maybe there is a little bug in the next swarm image release candidate. Otherwise, I'll get back to it in exactly 5 days.

**Comment by dongluochen** [comment]
I'm not aware of changes to data volume in Swarm 1.2.0-rc1. I'm surprised to see swarm:1.2.0-rc1 behaves differently from  swarm:latest (i.e., swarm:1.1.3 as for 3/24/16). If you can validate that, it may help us to locate the problem. 

You share a file from host to multiple VMs, then bind the file with multiple containers. I'm not sure how the updates are propagated. Timing could play a role in the sequence. Simultaneous access to the file may cause data corruption, IIRC.

**Comment by DarkFranX** [comment]
I will make a deeper investigation, try your suggestions and report back Tuesday!

**Comment by DarkFranX** [comment]
@dongluochen Good news, it is the 'iron/gcc:dev' container that does not update the file. Nothing to do with swarm. I don't want to ask too much, but do you have any idea of what is causing this?

**Comment by DarkFranX** [comment]
Problem solved using Apline 3.1 image. Thanks for your support! Closing.

**Comment by dongluochen** [comment]
@DarkFranX Thanks for your update!
